portable/ThirdParty/GCC/Posix/port.c

```diff
@@ -421,6 +422,14 @@ Thread_t *pxThread = pvParams;
 	/* Call the task's entry point. */
 	pxThread->pxCode( pxThread->pvParams );
 
+	/* The task has finished, we need to trigger a task switch here
+	 * to avoid exiting while all tasks are waiting to be signaled.
+	 * So we will mark the task as Dying here then delay the task
+	 * with any value to trigger a task switch where the task will
+	 * be suspended and exited. */
+	pxThread->xDying = pdTRUE;
+	vTaskDelay( portTASK_ENDED_DELAY );
```

Comment:
It will return to the same task when there are no other tasks at a higher or equal priority. So I get stuck again.

Reply:
The main issue is the delay
Why was it set to 500 ? would it work with 200? what about 100? etc
Can we make it so it doesn't wait at all?

Reply:
The delay can be with any value as I described it in the comment. The task actually will not wait at all. Once the task switching is in place, the task will exit as it is marked as "Dying". The purpose of the delay is to trigger an actual task switching and call `prvSwitchThread` with different task handles.

Reply:
Hmmm, I am still confused here. It seems like we are talking about the case where a FreeRTOS task function returns. By design when this happens the stack will be corrupted and the OS will crash, so what are we trying to do here?

I am trying to understand under which circumstances this code is helping us.

Reply:
A task should call vTaskDelete( NULL ) at its end.

Reply:
@RichardBarry The task resources will be freed in the idle task when the task deletes itself. This does not handle the following case:
1. Task (x) allocated memory to be used for a new higher priority static task (y).
2. Task (x) created a static task (y) and used the allocated memory from the previous step.
3. Task (y) scheduled then finished and deleted itself.
4. Scheduler returns to Task (x) which was waiting for Task (y) to finish.
5. Task (x) frees the allocated resources.
6. When The Idle task runs, it will try to free the resources of Task (y) and access a freed memory (and causes a Memory Corruption)

So I moved to delete Task (y) from the Task(x) and handle the task exit properly.
There is a workaround for this issue by adding `vTaskDelay` in Task (x) (before step 5) to give a chance for the idle task to run and free the resources of Task (y).

@cobusve Yes, I was trying to handle the case where the task function finishes. And It looks like it is not the proper way by design. I got confused because the task exit silently in Posix port without assertion or crashes. I will update the PR to just assert in this case to be easy to catch.

Reply:
@RichardBarry @cobusve  any suggestion on how to handle the above scenario?